File: tooling/ble_native/ios/BleAdvertiserPlugin.swift
import CoreBluetooth
import Flutter
import Foundation

/// Drop this file into ios/Runner/ after running `flutter create .` and
/// register it from AppDelegate. See tooling/ble_native/README.md.
///
/// Wraps CBPeripheralManager. Advertises our service UUID with the peer
/// id encoded as the local name — that's the one field iOS lets apps put
/// arbitrary text in while foregrounded.
///
/// Info.plist requires NSBluetoothAlwaysUsageDescription.
public class BleAdvertiserPlugin: NSObject, FlutterPlugin, CBPeripheralManagerDelegate, FlutterStreamHandler {

  private var methodChannel: FlutterMethodChannel?
  private var eventChannel: FlutterEventChannel?
  private var eventSink: FlutterEventSink?

  private var peripheral: CBPeripheralManager?
  private var pendingServiceUuid: String?
  private var pendingPeerId: String?
  private var isAdvertising = false

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = BleAdvertiserPlugin()
    let method = FlutterMethodChannel(name: "ubapp/ble_advertiser", binaryMessenger: registrar.messenger())
    let events = FlutterEventChannel(name: "ubapp/ble_advertiser/events", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: method)
    events.setStreamHandler(instance)
    instance.methodChannel = method
    instance.eventChannel = events
  }

  // MARK: - Method channel
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "isAvailable":
      result(CBCentralManager.authorization == .allowedAlways || CBCentralManager.authorization == .notDetermined)
    case "requestPermissions":
      // Triggers the permission prompt by instantiating the peripheral
      // manager. iOS doesn't have a separate request API.
      ensurePeripheral()
      result(true)
    case "start":
      guard let args = call.arguments as? [String: Any],
            let serviceUuid = args["serviceUuid"] as? String,
            let peerId = args["peerId"] as? String else {
        result(FlutterError(code: "bad_args", message: "serviceUuid and peerId required", details: nil))
        return
      }
      pendingServiceUuid = serviceUuid
      pendingPeerId = peerId
      ensurePeripheral()
      if peripheral?.state == .poweredOn {
        startAdvertising()
      } else {
        emit("starting", error: nil)
      }
      result(nil)
    case "stop":
      stopAdvertising()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Event channel
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }

  // MARK: - Peripheral manager
  private func ensurePeripheral() {
    if peripheral == nil {
      peripheral = CBPeripheralManager(delegate: self, queue: nil)
    }
  }

  private func startAdvertising() {
    guard let p = peripheral, let uuid = pendingServiceUuid, let peerId = pendingPeerId else { return }
    guard p.state == .poweredOn else {
      emit("starting", error: nil)
      return
    }
    if isAdvertising { p.stopAdvertising() }
    let serviceUuid = CBUUID(string: uuid)
    let data: [String: Any] = [
      CBAdvertisementDataServiceUUIDsKey: [serviceUuid],
      CBAdvertisementDataLocalNameKey: peerId,
    ]
    p.startAdvertising(data)
    isAdvertising = true
  }

  private func stopAdvertising() {
    peripheral?.stopAdvertising()
    isAdvertising = false
    emit("stopped", error: nil)
  }

  // MARK: - CBPeripheralManagerDelegate
  public func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    switch peripheral.state {
    case .poweredOn:
      if pendingServiceUuid != nil { startAdvertising() }
    case .poweredOff:
      emit("error", error: "Bluetooth is powered off")
    case .unauthorized:
      emit("error", error: "Bluetooth permission denied")
    case .unsupported:
      emit("unavailable", error: nil)
    case .resetting:
      emit("starting", error: nil)
    case .unknown:
      emit("starting", error: nil)
    @unknown default:
      emit("error", error: "Unknown state")
    }
  }

  public func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
    if let e = error {
      emit("error", error: e.localizedDescription)
    } else {
      emit("advertising", error: nil)
    }
  }

  private func emit(_ status: String, error: String?) {
    eventSink?(["status": status, "error": error as Any])
  }
}

File: tooling/ble_native/android/BleAdvertiserPlugin.kt
package com.example.ubapp

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.os.Build
import android.os.ParcelUuid
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.UUID

/**
 * Drop this file into the generated Android module after running
 * `flutter create .`. See tooling/ble_native/README.md for the wiring.
 *
 * Talks to BluetoothLeAdvertiser. Advertises a service UUID + the peer
 * id encoded in the local name field, which `flutter_blue_plus` reads as
 * `ScanResult.advertisementData.advName` on the scanner side.
 *
 * Permissions required (manifest):
 *   BLUETOOTH_ADVERTISE   (API 31+)
 *   BLUETOOTH_CONNECT     (API 31+ — required for some advertise paths)
 *   BLUETOOTH, BLUETOOTH_ADMIN  (API < 31)
 */
class BleAdvertiserPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private var context: Context? = null
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null

    private var advertiser: BluetoothLeAdvertiser? = null
    private var callback: AdvertiseCallback? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "ubapp/ble_advertiser").also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, "ubapp/ble_advertiser/events").also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        stopAdvertising()
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        methodChannel = null
        eventChannel = null
        context = null
    }

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink) { eventSink = sink }
    override fun onCancel(arguments: Any?) { eventSink = null }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(getAdapter()?.isMultipleAdvertisementSupported == true)
            "requestPermissions" -> {
                // Activity-bound permission requests live in MainActivity; the
                // plugin just reports whether it could obtain the advertiser.
                result.success(getAdvertiser() != null)
            }
            "start" -> {
                val serviceUuid = call.argument<String>("serviceUuid")
                val peerId = call.argument<String>("peerId")
                if (serviceUuid == null || peerId == null) {
                    result.error("bad_args", "serviceUuid and peerId required", null)
                    return
                }
                try {
                    startAdvertising(serviceUuid, peerId)
                    result.success(null)
                } catch (e: SecurityException) {
                    emit("error", e.message)
                    result.error("permission", e.message, null)
                } catch (e: Exception) {
                    emit("error", e.message)
                    result.error("start_failed", e.message, null)
                }
            }
            "stop" -> {
                stopAdvertising()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun getAdapter(): BluetoothAdapter? {
        val mgr = context?.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        return mgr?.adapter
    }

    private fun getAdvertiser(): BluetoothLeAdvertiser? {
        val a = getAdapter() ?: return null
        return a.bluetoothLeAdvertiser
    }

    private fun startAdvertising(serviceUuid: String, peerId: String) {
        stopAdvertising()
        val adv = getAdvertiser() ?: run {
            emit("unavailable", "BluetoothLeAdvertiser not available")
            return
        }
        advertiser = adv
        emit("starting", null)

        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
            .setConnectable(false)
            .setTimeout(0)
            .build()

        val parcel = ParcelUuid(UUID.fromString(serviceUuid))
        val data = AdvertiseData.Builder()
            .setIncludeDeviceName(false)
            .addServiceUuid(parcel)
            // Pack peer id into 16-byte ASCII service data; scanner pulls
            // either this or scan-response local name.
            .addServiceData(parcel, peerId.toByteArray(Charsets.UTF_8).take(16).toByteArray())
            .build()

        val scanResponse = AdvertiseData.Builder()
            .setIncludeDeviceName(true)
            .build()

        try {
            // Set the BluetoothAdapter local name to our peer id so iOS
            // scanners (which surface `peripheral.name`) can read it.
            getAdapter()?.let { adapter ->
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S || hasConnectPermission()) {
                    adapter.name = peerId
                }
            }
        } catch (_: SecurityException) {
            // ignore — advertising still works, scanner falls back to service data
        }

        callback = object : AdvertiseCallback() {
            override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
                emit("advertising", null)
            }
            override fun onStartFailure(errorCode: Int) {
                emit("error", "advertise failure code=$errorCode")
            }
        }
        adv.startAdvertising(settings, data, scanResponse, callback)
    }

    private fun stopAdvertising() {
        val adv = advertiser ?: return
        val cb = callback ?: return
        try {
            adv.stopAdvertising(cb)
        } catch (_: SecurityException) {}
        advertiser = null
        callback = null
        emit("stopped", null)
    }

    private fun hasConnectPermission(): Boolean {
        return try {
            val ctx = context ?: return false
            val pm = ctx.packageManager
            val info = pm.getPackageInfo(ctx.packageName, android.content.pm.PackageManager.GET_PERMISSIONS)
            info.requestedPermissions?.contains("android.permission.BLUETOOTH_CONNECT") == true
        } catch (_: Exception) {
            false
        }
    }

    private fun emit(status: String, error: String?) {
        eventSink?.success(mapOf("status" to status, "error" to error))
    }
}
